grades/api_test.py

```diff
@@ -0,0 +1,237 @@
+        refr.assert_called_once_with(self.user.social_auth.get(provider=EdxOrgOAuth2.name))
+        enr.assert_not_called()
+        cert.assert_called_once_with(self.user, ANY)
+        grades.assert_called_once_with(self.user, ANY)
+
+    @patch('grades.api._compute_grade_for_non_fa', autospec=True)
+    @patch('grades.api._compute_grade_for_fa', autospec=True)
+    def test_get_final_grade(self, fa, non_fa):
+        """
+        tests for get_final_grade function
+        """
+        cruns = (self.course_run1, self.course_run3,)
+
+        for crun in cruns:
+            api.get_final_grade(self.user, crun)
+
+        assert fa.called is True
+        assert fa.call_count == 1
+        call_arg = fa.call_args_list[0][0][0]
+        assert isinstance(call_arg, UserCachedRunData)
+        assert call_arg.edx_course_key == self.course_run1.edx_course_key
+
+        assert non_fa.called is True
+        assert non_fa.call_count == 1
+        call_arg = non_fa.call_args_list[0][0][0]
+        assert isinstance(call_arg, UserCachedRunData)
+        assert call_arg.edx_course_key == self.course_run3.edx_course_key
+
+    def test_get_users_final_grade_freeze(self):
+        """
+        tests for get_users_final_grade_freeze function
+        """
+        assert [user.pk for user in api.get_users_final_grade_freeze(self.course_run1)] == [self.user.pk]
+
+        # create another user and enrollment
+        other_user = UserFactory.create()
+        CachedEnrollmentFactory.create(user=other_user, course_run=self.course_run1)
+        assert sorted([user.pk for user in
+                       api.get_users_final_grade_freeze(self.course_run1)]) == sorted([self.user.pk, other_user.pk])
+
+        # add the user to the FinalGrade model as in progress
+        fg_status = FinalGrade.objects.create(
+            user=other_user, course_run=self.course_run1, status=FinalGradeStatus.PENDING, grade=0.0)
+        assert sorted([user.pk for user in
+                       api.get_users_final_grade_freeze(self.course_run1)]) == sorted([self.user.pk, other_user.pk])
+
+        # change the final grade status to complete
+        fg_status.status = FinalGradeStatus.COMPLETE
+        fg_status.save()
+        assert [user.pk for user in api.get_users_final_grade_freeze(self.course_run1)] == [self.user.pk]
+
+    @patch('grades.api.get_final_grade', autospec=True)
+    @patch('grades.api._refresh_cache_final_grade', autospec=True)
+    def test_freeze_user_final_grade_error1(self, mock_refr, mock_get_fg):
+        """
+        Test for freeze_user_final_grade function in case of problems with can_freeze_grades
+        """
+        # case not ready to be frozen
+        api.freeze_user_final_grade(self.user, self.course_run3)
+        assert mock_refr.called is False
+        assert mock_get_fg.called is False
+        assert FinalGrade.objects.filter(user=self.user, course_run=self.course_run3).exists() is False
```

Comment:
the error is because the the original function will terminate in an erroneous state (without raising anything).
I think `test_future_freeze_date` is more misleading because suggests that the function should work even for future courses